File: 백준/Gold/2268. 수들의 합 7/수들의 합 7.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def modify(p, val, tree, n) -> None:
  p += n
  tree[p] = val
  while p > 1:
    tree[p >> 1] = tree[p] + tree[p ^ 1]
    p >>= 1

def sumq(l, r, tree, n) -> int:
  ret = 0
  l, r = l + n, r + n
  while l < r:
    if l & 1:
      ret += tree[l]
      l += 1
    if r & 1:
      r -= 1
      ret += tree[r]
    l >>= 1
    r >>= 1
  return ret

def main() -> None:
  # input & init
  n, m = map(int, input().split())

  # 세그먼트 트리
  tree = [ 0 for _ in range(n * 2) ]

  # Query
  for cmd, a, b in (map(int, input().split()) for _ in range(m)):
    if cmd: modify(a-1, b, tree, n)
    else: 
      print(sumq(min(a, b)-1, max(a, b), tree, n))

sys.exit(main())

File: 백준/Gold/2268. 수들의 합 7/수들의 합 7.java
// Authored by : chjh2129
import java.io.*;
import java.util.*;
public class Main {
  
  static int n, m;
  static long[] tree;

  public static void modify(int p, long val){
    p += n;
    tree[p] = val;
    for(int i = p; i > 1; i >>= 1) tree[i >> 1] = tree[i] + tree[i ^ 1];
  }

  public static long sumQuery(int l, int r){
    long ret = 0;
    for(l += n, r += n; l < r; l >>= 1, r >>= 1){
      if((l & 1) == 1) ret += tree[l++];
      if((r & 1) == 1) ret += tree[--r];
    }
    return ret;
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    StringTokenizer st = new StringTokenizer(br.readLine());

    // input & init
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());

    tree = new long[n * 2];

    // Query
    while(m -- > 0){
      st = new StringTokenizer(br.readLine());
      int cmd = Integer.parseInt(st.nextToken());
      int a = Integer.parseInt(st.nextToken());
      long b = Long.parseLong(st.nextToken());

      if(cmd == 1) modify(a-1, b);
      else sb.append(sumQuery(Math.min(a, (int)b)-1, Math.max(a, (int)b)) + "\n");
    }

    // output
    System.out.println(sb.toString());
  }
}